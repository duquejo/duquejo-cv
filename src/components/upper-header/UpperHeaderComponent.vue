
<template>
  <div class="sticky lg:relative flex justify-between w-full p-2 lg:hidden items-center">
    <span class="p-4" />
    <NameTagComponent classes="before:bg-gray-100"/>
    <v-icon :name="menuIcon" scale="1.5" @click="emit('menu-click')" />
  </div>
</template>

<script setup lang="ts">
import { computed } from 'vue';
import NameTagComponent from '@/components/name-tag/NameTagComponent.vue';

interface Props {
  isOpen: boolean;
}

interface Emits {
  (e: 'menu-click'): void;
}

const props = defineProps<Props>();
const emit = defineEmits<Emits>();

const menuIcon = computed(() => props.isOpen ? 'co-hamburger-menu' : 'pr-times' );
</script>