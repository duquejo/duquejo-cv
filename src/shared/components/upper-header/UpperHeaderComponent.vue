
<template>
  <div class="flex justify-between w-full p-2 lg:hidden items-center">
    <span class="p-4" />
    <span class="relative w-[max-content] font-mono before:absolute before:inset-0 before:animate-typewriter before:bg-gray-100 before:opacity-80 after:absolute after:inset-0 after:w-[0.125em] after:animate-caret after:bg-black tracking-widest">@duquejo</span>
    <v-icon :name="menuIcon" scale="1.5" />
  </div>
</template>

<script setup lang="ts">
import { computed } from 'vue';
interface Props {
  isOpen: boolean;
}
const props = defineProps<Props>();
const menuIcon = computed(() => props.isOpen ? 'co-hamburger-menu' : 'pr-times' );
</script>