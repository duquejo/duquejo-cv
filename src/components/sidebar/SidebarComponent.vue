<template>
  <aside id="sidebar" class="transition-all duration-500 absolute flex flex-col items-center justify-between lg:px-8 lg:py-10 lg:my-2 lg:mx-2 rounded-lg bg-white lg:h-auto z-10 w-full lg:w-auto lg:relative lg:opacity-100 lg:left-0 h-screen" :class="classes">
    <UpperHeaderComponent :isOpen="isOpen" @click="emit('toggle')"/>
    <div class="flex flex-col justify-center items-center w-full lg:p-0 px-8 py-10 text-center text-base">
      <picture class="mb-5">
        <img class="rounded-full w-32 h-32 sm:w-40 sm:h-40 object-cover grayscale hover:grayscale-0 transition-all duration-200" src="../../assets/avatar-small.jpg" alt="José Duque" />
      </picture>
      <NameTagComponent classes="mb-1 before:bg-transparent text-gray-700 hidden lg:block" size="small"/>
      <span class="text-2xl font-semibold sm:text-xl">José Duque</span>
      <span class="text-gray-500">{{ t('general.role') }}</span>
      <div v-if="props.socialLinks?.length" class="sidebar-news--container">
        <i class="icon" v-for="link in props.socialLinks" :key="link.name">
          <a :href="link.url" target="_blank">
            <v-icon :name="link.icon" scale="1.3" class="fill-gray-400 hover:fill-yellow-500 transition-colors" />
          </a>
        </i>
      </div>
      <LangSelectComponent id="sidebar-language-selector" class="lg:hidden" />
    </div>
    <ul class="navigation text-right my-5 space-y-2 w-full lg:p-0 px-8 py-10">
      <li>
        <router-link :to="{ name: ROUTES.RESUME }" @click="emit('toggle')">
          {{ t('general.menu.resume') }}<span><v-icon class="ml-2" name="bi-chevron-right" scale="0.9" /></span>
        </router-link>
      </li>
      <li>
        <router-link :to="{ name: ROUTES.SERVICES }" @click="emit('toggle')">
          {{ t('general.menu.services') }}<span><v-icon class="ml-2" name="bi-chevron-right" scale="0.9" /></span>
        </router-link>
      </li>
    </ul>
  </aside>
</template>

<script setup lang="ts">
import { computed } from 'vue';
import { useI18n } from 'vue-i18n';

import { ROUTES } from '@/router';
import { SocialLinks } from '@/interfaces';

import UpperHeaderComponent from '@/components/upper-header/UpperHeaderComponent.vue';
import LangSelectComponent from '@/components/language-select/LangSelectComponent.vue';
import NameTagComponent from '@/components/name-tag/NameTagComponent.vue';

interface Props {
  isOpen: boolean;
  socialLinks?: SocialLinks[];
}

interface Emit {
  (e: 'toggle'): void;
}

const { t } = useI18n();
const props = defineProps<Props>();
const emit = defineEmits<Emit>();

const classes = computed(() => ({
  'opacity-0 left-full': props.isOpen,
  'opacity-100 left-0': ! props.isOpen,
}));

</script>