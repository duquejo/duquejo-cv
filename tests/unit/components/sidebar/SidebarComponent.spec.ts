import { describe, expect, it, vi, beforeEach } from 'vitest';
import { shallowMount, VueWrapper, RouterLinkStub } from '@vue/test-utils';

import SidebarComponent from '../../../../src/components/sidebar/SidebarComponent.vue';
import UpperHeaderComponent from '../../../../src/components/upper-header/UpperHeaderComponent.vue';
import createVueI18n from '../../../mocks/mock-i18n';

import { SocialLinksBuilder } from '../../../utils/SocialLinksBuilder';

describe('+ SidebarComponent.vue tests', () => {

  const i18n = createVueI18n('en');

  let wrapper: VueWrapper;

  const social_links = new SocialLinksBuilder()
    .withDefaultLinks()
    .build();

  const createWrapper = (
    isOpen = false, 
    socialLinks = social_links
  ) => shallowMount(SidebarComponent, {
    props: {
      isOpen,
      socialLinks
    },
    global: {
      plugins: [i18n],
      stubs: {
        RouterLink: RouterLinkStub,
        UpperHeaderComponent: UpperHeaderComponent,
      }
    },
  });

  beforeEach(async () => {
    vi.clearAllMocks();
  });

  it('Should match the snapshot', () => {
    wrapper = createWrapper();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Should match the snapshot if socialLinks are not given', () => {
    wrapper = createWrapper(false, []);
    expect(wrapper.html()).toMatchSnapshot();

    const sidebarContainer = wrapper.find('.sidebar-news-container');
    expect(sidebarContainer.exists()).toBeFalsy();
  });

  it('should render with the sidebar closed by default', () => {
    wrapper = createWrapper();
    const sidebar = wrapper.find('#sidebar');
    expect(sidebar.classes()).toContain('left-0');
    expect(sidebar.classes()).not.toContain('left-full');
  });

  it('should render with the sidebar open if isOpen is true', () => {
    wrapper = createWrapper(true);
    const sidebar = wrapper.find('#sidebar');
    expect(sidebar.classes()).toContain('left-full');
    expect(sidebar.classes()).not.toContain('left-0');
  });

  it('should emit "toggle" event when UpperHeaderComponent is clicked', async () => {
    const wrapper = createWrapper();
    const upperHeader = wrapper.findComponent(UpperHeaderComponent);
    await upperHeader.trigger('click');
    expect(wrapper.emitted()).toHaveProperty('toggle');
  });
});