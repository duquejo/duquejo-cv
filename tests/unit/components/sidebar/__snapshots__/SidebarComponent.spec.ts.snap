// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`+ SidebarComponent.vue tests > Should match the snapshot 1`] = `
"<aside id="sidebar" class="transition-all duration-500 absolute flex flex-col items-center justify-between lg:px-8 lg:py-10 lg:my-2 lg:mx-2 rounded-lg bg-white lg:h-auto z-10 w-full lg:w-auto lg:relative lg:opacity-100 lg:left-0 h-screen opacity-100 left-0" sociallinks="[object Object],[object Object]">
  <div class="flex justify-between w-full p-2 lg:hidden items-center"><span class="p-4"></span>
    <name-tag-component-stub classes="before:bg-gray-100" size="medium"></name-tag-component-stub>
    <v-icon name="pr-times" scale="1.5"></v-icon>
  </div>
  <div class="flex flex-col justify-center items-center w-full lg:p-0 px-8 py-10 text-center text-base">
    <picture class="mb-5 max-h-40 max-w-40 overflow-hidden rounded-full"><img class="sm:w-40 sm:h-40 object-cover hover:grayscale-0 transition-all duration-200 hover:scale-110 delay-200" src="/src/assets/avatar-small.webp" width="160" height="160" alt="José Duque"></picture>
    <name-tag-component-stub classes="mb-1 before:bg-transparent text-gray-700 hidden lg:block" size="small"></name-tag-component-stub><span class="text-2xl font-semibold sm:text-xl">José Duque</span><span class="text-gray-600">general.role</span><small class="border-t border-gray-200 pt-px mt-5 mb-0 lg:hidden">general.sidebar.network_links</small>
    <div class="flex flex-row justify-evenly lg:my-2 gap-x-3"><i class="icon"><a href="https://github.com/duquejo01" target="_blank" aria-label="general.sidebar.network_links_label github">
          <v-icon name="bi-github" scale="1.3" class="fill-gray-400 hover:fill-yellow-400 transition-colors"></v-icon>
        </a></i><i class="icon"><a href="https://www.linkedin.com/in/duquejo" target="_blank" aria-label="general.sidebar.network_links_label linkedln">
          <v-icon name="co-linkedin-in" scale="1.3" class="fill-gray-400 hover:fill-yellow-400 transition-colors"></v-icon>
        </a></i></div><small class="border-t border-gray-200 pt-px mt-3 mb-0 lg:hidden">general.sidebar.theme_options</small>
    <lang-select-component-stub id="sidebar-language-selector" class="lg:hidden"></lang-select-component-stub>
  </div>
  <navigation-links-stub></navigation-links-stub>
</aside>"
`;

exports[`+ SidebarComponent.vue tests > Should match the snapshot if socialLinks are not given 1`] = `
"<aside id="sidebar" class="transition-all duration-500 absolute flex flex-col items-center justify-between lg:px-8 lg:py-10 lg:my-2 lg:mx-2 rounded-lg bg-white lg:h-auto z-10 w-full lg:w-auto lg:relative lg:opacity-100 lg:left-0 h-screen opacity-100 left-0" sociallinks="">
  <div class="flex justify-between w-full p-2 lg:hidden items-center"><span class="p-4"></span>
    <name-tag-component-stub classes="before:bg-gray-100" size="medium"></name-tag-component-stub>
    <v-icon name="pr-times" scale="1.5"></v-icon>
  </div>
  <div class="flex flex-col justify-center items-center w-full lg:p-0 px-8 py-10 text-center text-base">
    <picture class="mb-5 max-h-40 max-w-40 overflow-hidden rounded-full"><img class="sm:w-40 sm:h-40 object-cover hover:grayscale-0 transition-all duration-200 hover:scale-110 delay-200" src="/src/assets/avatar-small.webp" width="160" height="160" alt="José Duque"></picture>
    <name-tag-component-stub classes="mb-1 before:bg-transparent text-gray-700 hidden lg:block" size="small"></name-tag-component-stub><span class="text-2xl font-semibold sm:text-xl">José Duque</span><span class="text-gray-600">general.role</span><small class="border-t border-gray-200 pt-px mt-5 mb-0 lg:hidden">general.sidebar.network_links</small>
    <div class="flex flex-row justify-evenly lg:my-2 gap-x-3"><i class="icon"><a href="https://github.com/duquejo01" target="_blank" aria-label="general.sidebar.network_links_label github">
          <v-icon name="bi-github" scale="1.3" class="fill-gray-400 hover:fill-yellow-400 transition-colors"></v-icon>
        </a></i><i class="icon"><a href="https://www.linkedin.com/in/duquejo" target="_blank" aria-label="general.sidebar.network_links_label linkedln">
          <v-icon name="co-linkedin-in" scale="1.3" class="fill-gray-400 hover:fill-yellow-400 transition-colors"></v-icon>
        </a></i></div><small class="border-t border-gray-200 pt-px mt-3 mb-0 lg:hidden">general.sidebar.theme_options</small>
    <lang-select-component-stub id="sidebar-language-selector" class="lg:hidden"></lang-select-component-stub>
  </div>
  <navigation-links-stub></navigation-links-stub>
</aside>"
`;
